### ISSUE TITLE
Race-conditions in  mainThreadScheduler ?

### ISSUE THREAD

**Issue by bobymicroby** [ReactiveX/RxAndroid#300]
Hello guys,

I have watched the presentation of Ray Ryan : [Where the Reactive Rubber Meets the Road](https://www.youtube.com/watch?v=va1d4MqLUGY) . Then I had a look at ```LooperScheduler```  and I have found  ```  handler.sendMessageDelayed(message, unit.toMillis(delayTime)); ```  as a way to schedule actions. So I guess there is a real probability of having some view objects out of scope in the subscribers?

Do you think that this should be stated in the documentation more explicitly ? 
This can cause really nasty bugs, that are going to be so hard to find, because 
most of the people out there have the firm believe that if they schedule on the mainThreadScheduler
everything is going to be fine.

P.S Do you think that using synchronous subjects as Ray suggests is the way to work around this issue ?

Bobby.

**Comment by JakeWharton** [comment]
> So I guess there is a real probability of having some view objects out of scope in the subscribers?

For asynchronous programming this is always true. RxJava provides unsubscribing from a `Subscription` to work around this so it should be a non-issue as long as you plug this into your view lifecycle.

> Do you think that this should be stated in the documentation more explicitly ? 

No. This is _always_ true when moving threads. RxJava is synchronous by default so once you start using schedulers to move threads you have to deal with the implications of multi-threaded programming that are present regardless of whether you're using Rx, Handler, Executors, or Threads.

> This can cause really nasty bugs, that are going to be so hard to find, because 
most of the people out there have the firm believe that if they schedule on the mainThreadScheduler
everything is going to be fine.

The implementation is no different than calling `runOnUiThread` or using an `AsyncTask` though.

> Do you think that using synchronous subjects as Ray suggests is the way to work around this issue ?

There is no issue. That presentation deals with communication between a presenter and the UI being synchronous instead of asynchronous by not re-applying schedulers when you're already on the main thread.

**Comment by JakeWharton** [comment]
TL;DR: There's no race condition, there's only normal multi-threaded programming considerations.

I'm going to close this unless there's some explicit action recommended here.

**Comment by bobymicroby** [comment]
So we just need to make sure we unsubscribe accorsing to the activity lifecycle and doublecheck if the views are reachable in the subsribers. Thanks for the detailed answer.

**Comment by JakeWharton** [comment]
You should never need to check for the view in the subscriber.

**Comment by bobymicroby** [comment]
Okay then. Thanka once more.

**Comment by felipecsl** [comment]
@bobymicroby take a look into libraries like RxLifecycle or RxGroups to aid with the lifecycle management

**Comment by bobymicroby** [comment]
@felipecsl Will do . Thanks again for the support guys .
